examples/dodo_prompt.rs

```diff
@@ -77,8 +82,73 @@ impl Completer for CommentCompleter {
     }
 }
 
+pub struct NewlineForBackslashAndOpenDelimRule;
+
+impl NewlineRule for NewlineForBackslashAndOpenDelimRule {
+    fn evaluate_on_newline(&self, buf: &Buffer) -> bool {
+        last_non_ws_char_was_not_backslash(buf) && check_balanced_delimiters(buf)
+    }
+}
+
+pub fn check_balanced_delimiters(buf: &Buffer) -> bool {
+    let buf_vec = buf.range_graphemes_all();
+    let mut stack = vec![];
```

Comment:
Can't this just be a counter instead of a vec, it looks like there is no need to allocate a vector for this based on it's use.

Reply:
uh, this code was not even correct. I spruced it up to be correct, and is the kind of thing that we could now use in sl-sh if we want. The old code didn't actually care if a ( was matched with another ), you could have matched it with a } or a ] and it would have worked. I also added the concept of being inside of a double quote so we were not trying to balance delimiters in user strings.

The thing I decided to not do, was refuse to evalute if we had this business

```
$> cargo run --example dodo_prompt history
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/examples/dodo_prompt history`
History file: history
[prompt]
% ls (
  ]]]]]]]]
  )
ENTERED: [ls (
]]]]]]]]
)]
[prompt]
%
```

basically if YOU open a delimiter, we'll make sure it has a matching close, but if you add closing delimiters that don't have corresponding opening ones... whatever.

My thought process is that it's just less intuitive. if you think something should have evaluates when you hit enter and it did not then the first assumption is that something was opened but not closed. not that something was closed that was never opened. 

Now that I think about it, maybe this idea of `strict balance` becomes optional? TBH I don't know if I'd want to enable it, but it does have some value.